src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -0,0 +1,320 @@
+  PeakIntegrator::PeakShapeMetrics PeakIntegrator::getPeakShapeMetrics() const { return psm_; }
+
+  void PeakIntegrator::getDefaultParameters(Param& params)
+  {
+    params.clear();
+    params.setValue("integration_type", "intensity_sum", "The integration technique to use in integratePeak() and estimateBackground().");
+    params.setValidStrings("integration_type", ListUtils::create<String>("intensity_sum,simpson,trapezoid"));
+    params.setValue("baseline_type", "base_to_base", "The baseline type to use in estimateBackground().");
+    params.setValidStrings("baseline_type", ListUtils::create<String>("base_to_base,vertical_division"));
+  }
+
+  void PeakIntegrator::updateMembers_()
+  {
+    integration_type_ = (String)param_.getValue("integration_type");
+    baseline_type_ = (String)param_.getValue("baseline_type");
+  }
+
+  template <typename PeakContainerT>
+  void PeakIntegrator::estimateBackground_(const PeakContainerT& p, const double& left, const double& right)
+  {
+    const double int_l = p.PosBegin(left)->getIntensity();
+    const double int_r = (p.PosEnd(right)-1)->getIntensity();
+    const double delta_int = int_r - int_l;
+    const double delta_pos = (p.PosEnd(right)-1)->getPos() - p.PosBegin(left)->getPos();
+    const double min_int_pos = int_r <= int_l ? (p.PosEnd(right)-1)->getPos() : p.PosBegin(left)->getPos();
+    const double delta_int_apex = std::fabs(delta_int) * std::fabs(min_int_pos - peak_apex_pos_) / delta_pos;
+    background_height_ = std::min(int_r, int_l) + delta_int_apex;
+    double background = 0.0;
+    if (baseline_type_ == "base_to_base")
+    {
+      if (integration_type_ == "trapezoid" || integration_type_ == "simpson")
+      {
+        // formula for calculating the background using the trapezoidal rule
+        // background = intensity_min*delta_pos + 0.5*delta_int*delta_pos;
+        background = delta_pos * (std::min(int_r, int_l) + 0.5 * std::fabs(delta_int));
+      }
+      else if (integration_type_ == "intensity_sum")
+      {
+        // calculate the background using the formula
+        // y = mx + b where x = rt or mz, m = slope, b = left intensity
+        // sign of delta_int will determine line direction
+        // background += delta_int / delta_pos * (it->getPos() - left) + int_l;
+        UInt n_points = 0;
+        for (auto it=p.PosBegin(left); it!=p.PosEnd(right); ++it, ++n_points)
+        {
+          background += it->getPos();
+        }
+        background = (background - n_points * p.PosBegin(left)->getPos()) * delta_int / delta_pos + n_points * int_l;
+      }
+    }
+    else if (baseline_type_ == "vertical_division")
+    {
+      if (integration_type_ == "trapezoid" || integration_type_ == "simpson")
+      {
+        background = delta_pos * std::min(int_r, int_l);
+      }
+      else if (integration_type_ == "intensity_sum")
+      {
+        UInt n_points = 0;
+        for (auto it=p.PosBegin(left); it!=p.PosEnd(right); ++it, ++n_points)
```

Comment:
this can be simpler written with std::distance
also: are you sure you are counting the number of points correctly?

Reply:
if not, please also check the other occurences of this type of coe

Reply:
I will use that `std::distance`, seems the method I should be using!
Yes I think I've been counting them correctly. I guess `std::distance` will help me finding out the truth :smile: